ci-operator/step-registry/cert-manager/upgrade/cert-manager-upgrade-commands.sh

```diff
@@ -0,0 +1,175 @@
+#!/bin/bash
+
+set -e
+set -u
+set -o pipefail
+
+timestamp() {
+    date -u --rfc-3339=seconds
+}
+
+if [ -f "${SHARED_DIR}/proxy-conf.sh" ] ; then
+    source "${SHARED_DIR}/proxy-conf.sh"
+    echo "proxy: ${SHARED_DIR}/proxy-conf.sh"
+fi
+
+# Check if the catalogsource is ready to use
+if [[ ! "$(oc get catalogsource $UPGRADE_CATSRC -n openshift-marketplace -o yaml)" =~ "lastObservedState: READY" ]]; then
+    echo "Skipping as the catalogsource $UPGRADE_CATSRC is either not existing or not ready."
+    exit 0
+fi
+
+# Define common variables
+SUB="openshift-cert-manager-operator"
+OPERATOR_NAMESPACE="cert-manager-operator"
+OPERAND_NAMESPACE="cert-manager"
+ISSUER="sanity-upgrade-selfsigned"
+CERTIFICATE="sanity-upgrade-selfsigned-cert"
+TEST_NAMESPACE="test-cert-manager-upgrade"
+INTERVAL=10
+
+# Prepare the sanity test data for post-upgrade check
+
+echo "# Creating a namespace from test usage."
+oc create ns $TEST_NAMESPACE
+
+echo "# Creating a self-signed issuer."
+ISSUER_MANIFEST=$(cat <<EOF
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: $ISSUER
+  namespace: $TEST_NAMESPACE
+spec:
+  selfSigned: {}
+EOF
+)
+oc create -f - <<< "${ISSUER_MANIFEST}"
+
+echo "# Waiting for the issuer to become ready."
+MAX_RETRY=12
+COUNTER=0
+while :;
+do
+    echo "Checking the issuer $ISSUER status for the #${COUNTER}-th time ..."
+    if [[ "$(oc get --no-headers issuer $ISSUER -n $TEST_NAMESPACE)" =~ True ]]; then
+        echo "[$(timestamp)] The issuer has become ready."
+        break
+    fi
+    ((++COUNTER))
+    if [[ $COUNTER -eq $MAX_RETRY ]]; then
+        echo "[$(timestamp)] The issuer status is still not ready after $((MAX_RETRY * INTERVAL)) seconds. Dumping status:"
+        oc get issuer $ISSUER -n $TEST_NAMESPACE -o=jsonpath='{.status}'
+        exit 1
+    fi
+    sleep $INTERVAL
+done
+
+# Perform the upgrade process
+
+echo -e "# Patching the subscription with catalogsource '$UPGRADE_CATSRC' and channel '$UPGRADE_CHANNEL'."
+oc patch subscription "${SUB}" -n $OPERATOR_NAMESPACE --type merge --patch '{"spec":{"source":"'"${UPGRADE_CATSRC}"'","channel":"'"${UPGRADE_CHANNEL}"'"}}'
+
+echo "# Waiting for the upgrade process to complete."
+while :;
+do
+    INSTALLED_CSV=$(oc get subscription $SUB -n $OPERATOR_NAMESPACE -o=jsonpath='{.status.installedCSV}' || true)
+    if [[ "$INSTALLED_CSV" == "${UPGRADE_CSV}" ]]; then
+        echo "[$(timestamp)] The CSV is upgraded to $UPGRADE_CSV. Listing all installplans during the upgrade process:"
+        oc get installplan -n $OPERATOR_NAMESPACE
+        break
+    fi
+
+    INSTALL_PLAN=$(oc get subscription $SUB -n $OPERATOR_NAMESPACE -o=jsonpath='{.status.installplan.name}' || true)
```

Comment:
Should we ensure to wait for the new installplan appear first? In automation, the time window between oc patch and oc get here can be very short.

Reply:
Make sense to me. I also notice it in previous rehearsals. To avoid nested wait loops, I ended up adding a simple [sleep](https://github.com/openshift/release/pull/55016/files#diff-e99d1f0764bcf0993487dbd439ebfb659a1f5ce51cd2678a5f9592951137f140R106) after checking the CSV status and before getting the next installplan.

From the latest [log](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_release/55016/rehearse-55016-periodic-ci-openshift-cert-manager-operator-master-upgrade-e2e-operator-upgrade/1818815688584728576/artifacts/e2e-operator-upgrade/cert-manager-upgrade/build-log.txt), it seems to work fine and doesn't increase much more waiting time in total. (~30s)

Reply:
> In automation, the time window between oc patch and oc get here can be very short.

The time window between oc patch sub ... and oc get here can be very short and easily cause flakes.

Even if for oc patch installplan ...., your short sleep can reduce flakes but do not avoid flakes.